---
[v2] net: phy: add MaxLinear CUX3410 10GBase-T PHY support

Add driver support for MaxLinear CUX3410 10GBase-T Ethernet PHY including
initialization, auto-negotiation, and link management.
---
v2: Add a remove callback to CUX3410 to resolve the kernel panic that occurs
    when removing the driver.
v1: Initial CUX3410 10GBase-T PHY driver support.
---

From 0839c533136f039ca63e72871d955e9a6c1e3a1f Mon Sep 17 00:00:00 2001
From: Mason Chang <mason-cw.chang@mediatek.com>
Date: Mon, 13 Oct 2025 15:42:49 +0800
Subject: [PATCH] net: phy: aquantia: add cux3410 support

---
 drivers/net/phy/aquantia/aquantia.h      |  2 +-
 drivers/net/phy/aquantia/aquantia_leds.c |  6 ++-
 drivers/net/phy/aquantia/aquantia_main.c | 55 ++++++++++++++++++++++++
 3 files changed, 61 insertions(+), 2 deletions(-)

--- a/drivers/net/phy/aquantia/aquantia.h
+++ b/drivers/net/phy/aquantia/aquantia.h
@@ -77,7 +77,7 @@
 #define VEND1_GLOBAL_LED_PROV_LINK100		BIT(5)
 #define VEND1_GLOBAL_LED_PROV_RX_ACT		BIT(3)
 #define VEND1_GLOBAL_LED_PROV_TX_ACT		BIT(2)
-#define VEND1_GLOBAL_LED_PROV_ACT_STRETCH	GENMASK(0, 1)
+#define VEND1_GLOBAL_LED_PROV_ACT_STRETCH	GENMASK(1, 0)
 
 #define VEND1_GLOBAL_LED_PROV_LINK_MASK		(VEND1_GLOBAL_LED_PROV_LINK100 | \
 						 VEND1_GLOBAL_LED_PROV_LINK1000 | \
--- a/drivers/net/phy/aquantia/aquantia_leds.c
+++ b/drivers/net/phy/aquantia/aquantia_leds.c
@@ -110,11 +110,15 @@ int aqr_phy_led_hw_control_set(struct ph
 	if (rules & BIT(TRIGGER_NETDEV_TX))
 		val |= VEND1_GLOBAL_LED_PROV_TX_ACT;
 
+	/* Stretch activity by 100ms */
+	val |= VEND1_GLOBAL_LED_PROV_ACT_STRETCH;
+
 	return phy_modify_mmd(phydev, MDIO_MMD_VEND1, AQR_LED_PROV(index),
 			      VEND1_GLOBAL_LED_PROV_LINK_MASK |
 			      VEND1_GLOBAL_LED_PROV_FORCE_ON |
 			      VEND1_GLOBAL_LED_PROV_RX_ACT |
-			      VEND1_GLOBAL_LED_PROV_TX_ACT, val);
+			      VEND1_GLOBAL_LED_PROV_TX_ACT |
+			      VEND1_GLOBAL_LED_PROV_ACT_STRETCH, val);
 }
 
 int aqr_phy_led_active_low_set(struct phy_device *phydev, int index, bool enable)
--- a/drivers/net/phy/aquantia/aquantia_main.c
+++ b/drivers/net/phy/aquantia/aquantia_main.c
@@ -34,6 +34,7 @@
 #define PHY_ID_AQR813	0x31c31cb2
 #define PHY_ID_AQR112C	0x03a1b790
 #define PHY_ID_AQR112R	0x31c31d12
+#define PHY_ID_CUX3410	0x31c31dd3
 
 #define MDIO_PHYXS_VEND_IF_STATUS		0xe812
 #define MDIO_PHYXS_VEND_IF_STATUS_TYPE_MASK	GENMASK(7, 3)
@@ -183,6 +184,31 @@ static void aqr107_get_stats(struct phy_
 	}
 }
 
+static int aqr107_config_led(struct phy_device *phydev)
+{
+	int err;
+
+	/* LED0 */
+	err = aqr_phy_led_hw_control_set(phydev, 0,
+					 BIT(TRIGGER_NETDEV_LINK_10000));
+	if (err < 0)
+		return err;
+
+	/* LED1 */
+	err = aqr_phy_led_hw_control_set(phydev, 1,
+					 BIT(TRIGGER_NETDEV_LINK_100) |
+					 BIT(TRIGGER_NETDEV_LINK_1000) |
+					 BIT(TRIGGER_NETDEV_LINK_2500) |
+					 BIT(TRIGGER_NETDEV_LINK_5000));
+	if (err < 0)
+		return err;
+
+	/* LED2 */
+	return aqr_phy_led_hw_control_set(phydev, 2,
+					 BIT(TRIGGER_NETDEV_TX) |
+					 BIT(TRIGGER_NETDEV_RX));
+}
+
 static int aqr_config_aneg(struct phy_device *phydev)
 {
 	bool changed = false;
@@ -628,6 +654,10 @@ static int aqr107_config_init(struct phy
 	if (ret)
 		return ret;
 
+	ret = aqr107_config_led(phydev);
+	if (ret)
+		return ret;
+
 	ret = aqr107_config_mdi(phydev);
 	if (ret)
 		return ret;
@@ -1293,6 +1323,31 @@ static struct phy_driver aqr_driver[] =
 	.get_strings	= aqr107_get_strings,
 	.get_stats	= aqr107_get_stats,
 },
+{
+	PHY_ID_MATCH_MODEL(PHY_ID_CUX3410),
+	.name		= "Aquantia CUX3410",
+	.probe		= aqr107_probe,
+	.remove		= aqr107_remove,
+	.get_rate_matching = aqr107_get_rate_matching,
+	.config_init	= aqr113c_config_init,
+	.config_aneg	= aqr_config_aneg,
+	.config_intr	= aqr_config_intr,
+	.handle_interrupt	= aqr_handle_interrupt,
+	.read_status	= aqr107_read_status,
+	.get_tunable	= aqr107_get_tunable,
+	.set_tunable	= aqr107_set_tunable,
+	.suspend	= aqr107_suspend,
+	.resume		= aqr107_resume,
+	.get_sset_count = aqr107_get_sset_count,
+	.get_strings	= aqr107_get_strings,
+	.get_stats	= aqr107_get_stats,
+	.link_change_notify = aqr107_link_change_notify,
+	.led_brightness_set = aqr_phy_led_brightness_set,
+	.led_hw_is_supported = aqr_phy_led_hw_is_supported,
+	.led_hw_control_set = aqr_phy_led_hw_control_set,
+	.led_hw_control_get = aqr_phy_led_hw_control_get,
+	.led_polarity_set = aqr_phy_led_polarity_set,
+},
 };
 
 module_phy_driver(aqr_driver);
@@ -1316,6 +1371,7 @@ static const struct mdio_device_id __may
 	{ PHY_ID_MATCH_MODEL(PHY_ID_AQR813) },
 	{ PHY_ID_MATCH_MODEL(PHY_ID_AQR112C) },
 	{ PHY_ID_MATCH_MODEL(PHY_ID_AQR112R) },
+	{ PHY_ID_MATCH_MODEL(PHY_ID_CUX3410) },
 	{ }
 };
 
